Angular: Remove cached NgModules and introduce a global queue during bootstrapping

Marklb opened the pull request:
Closes #24813

<!-- If your PR is related to an issue, provide the number(s) above; if it resolves multiple issues, be sure to break them up (e.g. "closes #1000, closes #1001"). -->

<!--

Thank you for contributing to Storybook! Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `main` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/react/contribute/how-to-contribute

-->


## What I did

<!-- Briefly describe what your PR does -->

Problem: Caching the NgModule per component, causes the NgModule generated for the first story to get used for the following stories, but they may not have generated the same NgModule.

### Explanation:

Assume the following is the CSF file:
```ts
export default {
  component: FooComponent,
}

export const Primary = { }
export const Secondary = {
  moduleMetadata: {
    declarations: [ BarComponent ],
  },
}
```

In the renderer, the following is basically what the NgModule would be for each story:
```ts
// Primary
@NgModule({
  declarations: [ FooComponent ],
})
class StorybookModule {}

// Secondary
@NgModule({
  declarations: [ FooComponent, BarComponent ],
})
class StorybookModule {}
```

When using inline rendering to render both stories in a single document, Storybook attempts to render each story, without waiting on the previous to complete. The problem is that Angular only allows a component to be declared by a single NgModule, so we have to clear the compiled component cache after each story's app is initialized.

The solution being used currently, would cache the NgModule by the `component` defined in `Meta`. That avoids the error, but breaks the "Secondary" story, because `BarComponent` is not declared in the NgModule that was cached when the "Primary" story rendered (*Assuming the "Primary" story rendered first.*).

I don't know if the way inline stories render changed and used to wait for completion of each story or if it just wasn't fully implemented as expected. `AbstractRenderer` already has a method that calls `ɵresetCompiledComponents` and both `CanvasRenderer` and `DocsRenderer` call it after a full render. The problem is two render methods can be running in parallel, which makes the reset not work correctly.

To fix `ɵresetCompiledComponents` not getting called correctly, a queue is introduced to the `AbstractRenderer` that prevents calling `bootstrapApplication` in parallel. This may have an effect on performance, if a particular story is slow to initialize, but I don't know of another way to avoid conflicts between each Angular application's compiling.

## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [x] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

To see the problem with caching the NgModule, 
1. Run a sandbox for template, e.g. `yarn task --task sandbox --start-from auto --template angular-cli/default-ts`
2. In the sandbox file `stories/core/decorators/componentWrapperDecorator/decorators.stories.ts`, add `tags: ['autodocs']` to the default export `Meta`.
3. Open Storybook in your browser
4. Access "frameworks > angular > core > decorators > componentWrapperDecorator > decorators > Docs" stories. Before this change, all the stories render something, but any of the stories declaring `ParentComponent` threw an error, because the first story did not declare `ParentComponent`. With this change all should render correctly.

To see the reason for adding the lock, instead of just removing the NgModule caching, you can do the same steps, but comment out `await acquireLock(selector);` in the the `bootstrapLock` function in `code/frameworks/angular/src/client/angular-beta/utils/BootstrapLock.ts`. Commenting that line would disable the lock and cause the *multiple modules declaring component* error that the caching was supposed to fix.

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

<!-- CANARY_RELEASE_SECTION -->

Marklb committed:
Remove cached NgModules and introduce a global lock during bootstrapping

Marklb committed:
Add tests and make bootstrapLock more generic

Marklb committed:
Merge branch 'next' into marklb/fix-multi-ng-app-rendering

valentinpalkovic committed:
Merge remote-tracking branch 'origin/next' into pr/Marklb/24982

valentinpalkovic committed:
Use a queue instead of loops and timeouts

valentinpalkovic commented:
@Marklb Thank you so much for figuring out the root cause of the issue. This was the missing peace. I am so grateful that you dived deep into it! 

I have thought about your solution, which works great. The only thing might be that renderings might happen in no particular order. I have pushed a different approach using queues. WDYT?

valentinpalkovic committed:
Reset compiled components after bootstrapping

valentinpalkovic commented:
No pressure! I will wait for your proper feedback. I also changed when the compiled modules got reset in the latest commit because otherwise, your newly written test case of mounting two angular components at once would fail.

Marklb commented on code/frameworks/angular/src/client/angular-beta/utils/BootstrapLock.ts:
```diff
@@ -0,0 +1,50 @@
+    const bootstrappingFn = queue.shift();
+    if (bootstrappingFn) {
+      await bootstrappingFn();
+      await resetCompiledComponents();
```
I haven't tested to see if it would work, but do you think it would make more sense to call this before bootstrapping or keep it after?

The following are my two thoughts:
1. By keeping it after, there may be more consistent usage.
2. If before, it should still prevent conflict with a previous render, but allow usage of not resetting. I don't know if `resetCompiledComponents` can have an effect on lazy loaded modules that aren't part of the initial render. I assume most situations that render multiple stories are dealing more with a documentation scenario, but if `resetCompiledComponents` is hiding an error that would typically happen, would it be better to not call it after? An isolated render in Canvas view could then expose that error, if possible. Of course, then we run into a situation of potential render inconsistency between Canvas and Docs.

valentinpalkovic replied:
I agree with your thoughts. I guess we can try it out to place it before? Technically, this might be a breaking change in behavior, but since we have started to work on 8.0.0, we will not release another minor release of Storybook 7.x. I would vote to turn around the logic, as you mentioned.

valentinpalkovic replied:
@Marklb Keeping the reset before the actual rendering was [causing issues](https://www.chromatic.com/build?appId=6322ce6af69825592bbb28fc&number=8400). In particular, registered modules like the `CommonModule weren't available. This has lead to styles not being applies since the CommonModule is necessary for the `styles` attribute being applied.

Marklb committed:
Merge branch 'next' into marklb/fix-multi-ng-app-rendering

Marklb committed:
Update test and names to refer to "queue" instead of "lock"

Marklb commented:
I updated a test to be more thorough, adjusted the names and wording from "lock" to "queue", and added back the comment explaining why the queue is necessary.

Other than getting your opinion on the questions I added about the `resetCompiledComponents` move, the changes look good to me.

valentinpalkovic commented on code/frameworks/angular/src/client/angular-beta/utils/BootstrapLock.ts:
```diff
@@ -0,0 +1,50 @@
+    const bootstrappingFn = queue.shift();
+    if (bootstrappingFn) {
+      await bootstrappingFn();
+      await resetCompiledComponents();
```
> By moving this here, doesn't it prevent anyone from having the option of implementing a renderer that doesn't reset compiled components?

You're right. In `@storybookjs/testing-angular,` I haven't found where you are inheriting from `AbstractRenderer,` though. I guess you mentioned, in general, that people might use it to implement a custom renderer, which inherits from `AbstractRenderer`. I used the GitHub search, and I was not able to find any custom renderer inheriting `AbstractRenderer`. Could you find some to figure out whether this is a real problem? But I would guess we can avoid calling `resetCompiledComponents` in the lock function. Do you have an idea how we might avoid it?

Marklb replied:
I am not sure how to best avoid it, at the moment. That can maybe wait for another PR, if necessary.

When I mentioned `@storybookjs/testing-angular`, I was just using it as an example, but forgot the renderer in that package isn't actually a render and doesn't extend `AbstractRenderer`. It is basically the renderer, without the bootstrapping part, because I create a component that can render the story and let the user's chosen test framework handle using it. So, this actually doesn't apply to it.

valentinpalkovic committed:
Reset compiled components before bootstraping

valentinpalkovic committed:
Merge branch 'next' into marklb/fix-multi-ng-app-rendering

valentinpalkovic committed:
Merge branch 'next' into marklb/fix-multi-ng-app-rendering

valentinpalkovic commented:
Hi @Marklb 

Could you try to resolve the conflicts? Then, I will give CI another try. If it still fails, I will take a look!

Marklb committed:
Merge branch 'next' into marklb/fix-multi-ng-app-rendering

Marklb committed:
Merge branch 'next' into marklb/fix-multi-ng-app-rendering

Marklb commented:
@valentinpalkovic The conflicts should be fixed now.

valentinpalkovic commented:
Great! I will take a look tomorrow at why CI is failing

valentinpalkovic committed:
Merge branch 'next' into marklb/fix-multi-ng-app-rendering

Marklb committed:
Merge branch 'next' into marklb/fix-multi-ng-app-rendering

Marklb committed:
Migrate tests to vitest

Marklb commented:
I tried to update the tests for vitest, but I am not familiar with vitest, yet. The `BootstrapQueue` tests sometimes pass and sometimes fail. After debugging, it seems that the tests are running in parallel, without being isolated. I am not positive that is what happens, but stepping through breakpoints, once a test does something async the next test starts then the next sync action will go back to a breakpoint in the first test.

I am just assuming that is the problem and maybe I need to add something to stop that or that isn't even the problem, but I will look into it later, when I have more time to dig into it.

valentinpalkovic commented:
Thanks for taking care of it! I try to support you as soon as I have some free resources. 

But maybe, first of all, it would make sense to understand why the UI Test in Chromatic fails.

Regarding the unit test: Can we run the whole test suite and its tasks sequentially? Maybe Vitest offers some options for doing so exclusively for a particular test suite.

valentinpalkovic committed:
Merge branch 'next' into marklb/fix-multi-ng-app-rendering

valentinpalkovic committed:
Merge remote-tracking branch 'origin/next' into pr/Marklb/24982

valentinpalkovic committed:
Fix flaky Angular tests

valentinpalkovic committed:
Refactor BootstrapQueue.ts to improve code execution order
